bin/main.ml

```diff
@@ -658,75 +658,87 @@ let check_path contexts =
+        (1, true)
+    in
+    let s = String.sub s ~pos ~len:(String.length s - pos) in
+    let path = Path.relative Path.root (prefix_target common s) in
+    check_path path;
+    if Path.is_root path then
+      die "@@ on the command line must be followed by a valid alias name"
+    else if not (Path.is_managed path) then
+      die "@@ on the command line must be followed by a relative path"
+    else
+      Ok [if is_rec then Alias_rec path else Alias path]
+  end else begin
+    let path = Path.relative Path.root (prefix_target common s) in
+    resolve_path path ~setup
+  end
+
+let log_targets ~log targets =
+  List.iter targets ~f:(function
+    | File path ->
+      Log.info log @@ "- " ^ (Path.to_string path)
+    | Alias path ->
+      Log.info log @@ "- alias " ^
+                      (Path.to_string_maybe_quoted path)
+    | Alias_rec path ->
+      Log.info log @@ "- recursive alias " ^
+                      (Path.to_string_maybe_quoted path));
+  flush stdout
+
+let resolve_targets_mixed ~log common (setup : Main.setup) user_targets =
   match user_targets with
   | [] -> []
   | _ ->
-    let check_path = check_path setup.contexts in
     let targets =
-      List.map user_targets ~f:(fun s ->
-        if String.is_prefix s ~prefix:"@" then begin
-          let pos, is_rec =
-            if String.length s >= 2 && s.[1] = '@' then
-              (2, false)
-            else
-              (1, true)
-          in
-          let s = String.sub s ~pos ~len:(String.length s - pos) in
-          let path = Path.relative Path.root (prefix_target common s) in
-          check_path path;
-          if Path.is_root path then
-            die "@@ on the command line must be followed by a valid alias name"
-          else if not (Path.is_managed path) then
-            die "@@ on the command line must be followed by a relative path"
-          else
-            Ok [if is_rec then Alias_rec path else Alias path]
-        end else begin
-          let path = Path.relative Path.root (prefix_target common s) in
-          check_path path;
-          let can't_build path =
-            Error (path, target_hint setup path);
-          in
-          if not (Path.is_managed path) then
-            Ok [File path]
-          else if Path.is_in_build_dir path then begin
-            if Build_system.is_target setup.build_system path then
-              Ok [File path]
-            else
-              can't_build path
-          end else
-            match
-              List.filter_map setup.contexts ~f:(fun ctx ->
-                let path = Path.append ctx.Context.build_dir path in
-                if Build_system.is_target setup.build_system path then
-                  Some (File path)
-                else
-                  None)
-            with
-            | [] -> can't_build path
-            | l  -> Ok l
-        end
-      )
-    in
+      List.map user_targets ~f:(function
+        | `String s -> resolve_target common ~setup s
```

Comment:
Could you use a plain variant here? I find polymorphic variants more error prone and they provide no real benefit here